theories/datatypes/List.ec

```diff
@@ -1818,6 +1819,113 @@ lemma mapiP x0 (f : int -> 'a -> 'b) (s : 'a list) y :
     exists n, (0 <= n && n < size s) /\ y = f n (nth x0 s n).
 proof. exact: mapi_recP. qed.
 
+(* -------------------------------------------------------------------- *)
+(*                          Element Replacement                         *)
+(* -------------------------------------------------------------------- *)
+op put (s : 'a list) (n : int) (x : 'a) =
+  if (0 <= n < size s)
+  then take n s ++ (x :: drop (n + 1) s)
+  else s.
+
+lemma put_empty (n : int) (x : 'a) :
+  put [] n x = [].
+proof. by smt(). qed.
+
+lemma size_put (s : 'a list) (n : int) (x : 'a) :
+  size (put s n x) = size s.
+proof.
+rewrite /put; case (0 <= n < size s) => [[ge0_n lts_n] | //].
+by rewrite size_cat size_take // lts_n /= size_drop /#.
+qed.
+
+lemma put_inrng (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => put s n x = take n s ++ (x :: drop (n + 1) s).
+proof. by rewrite /put => ->. qed. 
+
+lemma put_outrng (s : 'a list) (n : int) (x : 'a) :
+  !(0 <= n < size s) => put s n x = s.
+proof. by rewrite /put => ->. qed. 
+
+lemma put_cons_0 (s : 'a list) (x y : 'a) :
+  put (y :: s) 0 x = x :: s.
+proof. by rewrite /put /= drop0 addrC ltzS size_ge0. qed. 
+  
+lemma put_cons (s : 'a list) (n : int) (x y : 'a) :
+  1 <= n => put (y :: s) n x = y :: (put s (n - 1) x).
+proof. move => ge1_n @/put /#. qed.
+
+lemma put_cat_inrngfirst (s1 s2 : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s1 => put (s1 ++ s2) n x = put s1 n x ++ s2.
+proof.
+move => [ge0_n lts1_n]; rewrite /put size_cat ge0_n lts1_n /=.
+have ->: n < size s1 + size s2.
```

Comment:
We tend to prefer `/#` for pure arithmetic goals. they are fas and lead to shorter proofs. (Here, you might need `smt(size_ge0)`)